File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;
        HashSet<ArrayList> set = new HashSet<>();
        
        for (int i=0; i<grid.length; i++){
            for (int j=0; j<grid[0].length; j++){
                ArrayList<String> tmp = new ArrayList<>();
                tmp.add(Integer.toString(i));
                tmp.add(Integer.toString(j));
                if (!set.contains(tmp) && grid[i][j] == '1'){
                    count++;
                    dfs(i, j, set, grid);
                }
            }
        }
        return count;
    }
    
    public void dfs(int i, int j, HashSet<ArrayList> set, char[][] grid){
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == '0' || set.contains(new ArrayList<>(List.of(Integer.toString(i), Integer.toString(j))))) return;
        
        ArrayList<String> tmp = new ArrayList<>();
        tmp.add(Integer.toString(i));
        tmp.add(Integer.toString(j));
        set.add(tmp);
        
        dfs(i, j+1, set, grid);
        dfs(i, j-1, set, grid);
        dfs(i+1, j, set, grid);
        dfs(i-1, j, set, grid);
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        self.seen = set()
        
        def dfs(i, j):
            if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or grid[i][j] == "0":
                return
            if (i,j) in self.seen:
                return
            
            self.seen.add((i,j))
            dfs(i, j+1)
            dfs(i, j-1)
            dfs(i+1, j)
            dfs(i-1, j)
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1" and (i,j) not in self.seen:
                    count += 1
                    dfs(i,j)
        return count